src/SegmentedComputation.chpl

```diff
@@ -0,0 +1,89 @@
+module SegmentedComputation {
+  use CommAggregation;
+  use SipHash;
+
+  proc computeSegmentOwnership(segments: [?D] int, vD) {
+    // Locale that owns each segment's bytes
+    var startLocales: [D] int;
+    const low = vD.low;
+    const size = vD.size;
+    // Vector op on segments, should be fast
+    forall (sl, seg) in zip(startLocales, segments) {
+      // which locale do my bytes start on?
+      sl = ((seg - low)*numLocales) / size;
+    }
+
+    // Index of first segment each locale owns bytes for
+    var startSegInds: [LocaleSpace] int;
+    // Mark true where owning locale changes
+    const change = [(i, sl) in zip(D, startLocales)] if (i == D.low) then true else (sl != startLocales[i-1]);
+    // Wherever the owning locale increments, record that as first segment for the locale
+    forall (c, sl, i) in zip(change, startLocales, D) {
+      if c {
+        startSegInds[sl] = i;
+      }
+    }
+
+    // Number of segments each locale owns bytes for
+    var numSegs: [LocaleSpace] int;
+    // small number of iterations, no comms
+    for l in Locales {
+      if (l.id == numLocales - 1) {
+        numSegs[l.id] = D.size - startSegInds[l.id];
+      } else {
+        numSegs[l.id] = startSegInds[l.id + 1] - startSegInds[l.id];
+      }
+    }
+
+    var lengths: [D] int;
+    forall (l, s, i) in zip(lengths, segments, D) {
+      if (i == D.high) {
+        l = vD.size - s;
+      } else {
+        l = segments[i+1] - s;
+      }
+    }
+
+    return (startSegInds, numSegs, lengths);
+  }
+
+  enum SegFunction {
+    SipHash128,
+  }
+  
+  proc computeOnSegments(segments: [?D] int, values: [?vD] ?t, param function: SegFunction, type retType) throws {
+    var res: [D] retType;
+    if (D.size == 0) {
+      return res;
+    }
+
+    const (startSegInds, numSegs, lengths) = computeSegmentOwnership(segments, vD);
+    
+    // Start task parallelism
+    coforall loc in Locales {
+      on loc {
+        const myFirstSegIdx = startSegInds[loc.id];
+        const myNumSegs = numSegs[loc.id];
+        const mySegInds = {myFirstSegIdx..#myNumSegs};
+        // Segment offsets whose bytes are owned by loc
+        // Block transfer -- should use aggregation?
+        const mySegs = segments[mySegInds];
```

Comment:
@bradcray -- this is doing something like:

```chpl
const DR: [{start..#length}] int = BlockDist[{start..#length}];
```

Where we'll only be pulling from 1 or 2 remote locales. Surprisingly this results in privatizing something today, but I'm not sure why. Does that makes sense to you? Switching to:

```chpl
var DR: [{start..#length}] int = 0;
DR = BlockDist[{start..#length}];
```

gets rid of the privatization, but is uglier. 

Using `-schpl_serializeSlices` also gets rid of the privatization, but for this I'm more curious what we're even privatizing here and why.

Reply:
Not sure what I was thinking here -- of course slicing results in privatization today and that's why serializing slices is a benefit (I got myself confused here -- I had an older version that changed to split declaration and assignment and that got rid of privatization, but that was due to some other changes I had in the Chapel repo.)

@reuster986 today this slice assignment is fairly expensive due to the eager privatization (will be solved by https://github.com/chapel-lang/chapel/pull/18911, which I think we're hoping to get in this release) and would be further optimized by fixing https://github.com/chapel-lang/chapel/issues/14479. With hacks for both of those in place, this assignment gets much faster for me.

In the meantime I think the following aggregated version will be faster:

```chpl
var mySegs, myLens: [mySegInds] int;
forall i in mySegInds with (var agg = new SrcAggregator(int)) {
  agg.copy(mySegs[i], segments[i]);
  agg.copy(myLens[i], lengths[i]);
}
```

With my hacky optimizations in place the slice-assignment is faster than the aggregated version so I'm not sure which version we want to go with for right now. I might merge with what you have now, and I'll follow up with a PR to switch to aggregation after we get some performance data, and then open up an issue to ensure we revert that change when slice-serialization is enabled by default.

Reply:
Thanks! I updated to use aggregators for this.